src/urlpattern_constructor_string_parser.cpp

```diff
@@ -0,0 +1,579 @@
+        // change state given parser, "port", and 1.
+        else if (parser.is_port_prefix() &&
+                 parser.hostname_ipv6_bracket_depth == 0) {
+          parser.change_state(PARSER_STATE::PORT, 1);
+        }
+        // 4. Else if the result of running is a pathname start given parser
+        // is true, then run change state given parser, "pathname", and 0.
+        if (parser.is_pathname_start()) {
+          parser.change_state(PARSER_STATE::PATHNAME, 0);
+        }
+        // 5. Else if the result of running is a search prefix given parser is
+        // true, then run change state given parser, "search", and 1.
+        if (parser.is_search_prefix()) {
+          parser.change_state(PARSER_STATE::SEARCH, 1);
+        }
+        // 6. Else if the result of running is a hash prefix given parser is
+        // true, then run change state given parser, "hash", and 1.
+        if (parser.is_hash_prefix()) {
+          parser.change_state(PARSER_STATE::HASH, 1);
+        }
+        break;
+      }
+      case PARSER_STATE::PORT: {
+        // 1. If the result of running is a pathname start given parser is
+        // true, then run change state given parser, "pathname", and 0.
+        if (parser.is_pathname_start()) {
+          parser.change_state(PARSER_STATE::PATHNAME, 0);
+        }
+        // 2. Else if the result of running is a search prefix given parser is
+        // true, then run change state given parser, "search", and 1.
+        else if (parser.is_search_prefix()) {
+          parser.change_state(PARSER_STATE::SEARCH, 1);
+        }
+        // 3. Else if the result of running is a hash prefix given parser is
+        // true, then run change state given parser, "hash", and 1.
+        if (parser.is_hash_prefix()) {
+          parser.change_state(PARSER_STATE::HASH, 1);
+        }
+      }
+      case PARSER_STATE::PATHNAME: {
+        // 1. If the result of running is a search prefix given parser is
+        // true, then run change state given parser, "search", and 1.
+        if (parser.is_search_prefix()) {
+          parser.change_state(PARSER_STATE::SEARCH, 1);
+        }
+        // 2. Else if the result of running is a hash prefix given parser is
+        // true, then run change state given parser, "hash", and 1.
+        else if (parser.is_hash_prefix()) {
+          parser.change_state(PARSER_STATE::HASH, 1);
+        }
+        break;
+      }
+      case PARSER_STATE::SEARCH: {
+        // 1. If the result of running is a hash prefix given parser is true,
+        // then run change state given parser, "hash", and 1.
+        if (parser.is_hash_prefix()) {
+          parser.change_state(PARSER_STATE::HASH, 1);
+        }
+        break;
+      }
+      case PARSER_STATE::HASH: {
+        // do nothing;
+        break;
+      }
+      case PARSER_STATE::DONE: {
+        // 1. Assert: This step is never reached.
+        assert(false);
```

Comment:
We have ada development asserts that is only enabled on development builds. @lemire write a good blog post about this: https://lemire.me/blog/2023/03/15/runtime-asserts-are-not-free/?amp